llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp

```diff
@@ -996,10 +996,32 @@ bool Vectorizer::isSafeToMove(
   LLVM_DEBUG(dbgs() << "LSV: isSafeToMove(" << *ChainElem << " -> "
                     << *ChainBegin << ")\n");
 
-  assert(isa<LoadInst>(ChainElem) == IsLoadChain);
+  assert(isa<LoadInst>(ChainElem) == IsLoadChain &&
+         isa<LoadInst>(ChainBegin) == IsLoadChain);
+
   if (ChainElem == ChainBegin)
     return true;
 
+  if constexpr (IsLoadChain) {
+    // If ChainElem depends on ChainBegin, they're not safe to reorder.
+    SmallVector<Instruction *, 8> Worklist;
+    Worklist.emplace_back(ChainElem);
+    while (!Worklist.empty()) {
+      Instruction *I = Worklist.pop_back_val();
+      for (Use &O : I->operands()) {
+        if (isa<PHINode>(O))
+          continue;
+        if (auto *J = dyn_cast<Instruction>(O)) {
+          if (J == ChainBegin) {
```

Comment:
Wouldn't this be an issue if this were any of the instructions in the chain, not just the beginning?

Reply:
Hm, I originally had this code in the loop below, but had to move it here due to the `std::next` in `BBIt`. Will think about how we can check everything in the chain in elegantly: unfortunately, the function is just passed two `Instruction *`, instead of a first-class chain.

Reply:
How do the other vectorizers, particularly SLP deal with this?

Reply:
I've barely looked at SLP myself, but I think it works with a first-class tree. I think the relevant function to look at is `canReorderOperands`.

Reply:
Actually, the code is correct, because `isSafeToMove` gets called in a loop from `std::next(ChainBegin)` to `ChainEnd`, so we are indeed checking all elements of the chain. I was wondering why we didn't pass a first-class chain, and found the answer.